CONTRIBUTING.md

```diff
@@ -109,48 +109,356 @@ Check out [building instructions](https://github.com/prestodb/presto/tree/master
+    
+  ```java
+  time0Bto100KB.add(nanos, TimeUnit.NANOSECONDS);
+  ```
+    
+  * But static import it and use NANOSECONDS in the code.
+    
+  ```java
+  import static java.util.concurrent.TimeUnit.NANOSECONDS;
+  ...
+  time0Bto100KB.add(nanos, NANOSECONDS);
+  ```
+    
+  * Prefer Immutable collections in Guava when possible. For example, instead of using
+    
+  ```java
+  expressions.stream()
+             .map(OriginalExpressionUtils::castToExpression)
+             .collect(Collectors.toList())
+  ```
+    
+  * Use ```toImmutableList()```
+    
+  ```java
+  import com.google.common.collect.ImmutableList;
+  ...
+  import static com.google.common.collect.ImmutableList.toImmutableList;
+  ...
+  expressions.stream()
+             .map(OriginalExpressionUtils::castToExpression)
+             .collect(toImmutableList())
+  ```
+
+* **Method argument validation**
+    * Usually needed for constructors
+
+      ```java
+      protected SqlScalarFunction(Signature signature)
+      {
+          this.signature = requireNonNull(signature, "signature is null");
+          checkArgument(signature.getKind() == SCALAR, "function kind must be SCALAR");
+      }
+      ```
+
+* **Use proper annotations**
+    * @Nullable
+    * @VisibleForTesting
+    * @Override
+    * @Experimental
+* **Use of Optional instead of bare metal nullable objects when appropriate**
+    * public method parameters
+    * Performance NON-CRITICAL path
+    * Creating Optional objects is not free. If the function call is in a critical loop, do NOT use Optional parameters.
+* **Alphabetize**
+    * Sections in documentation
+    * Methods
+    * Variables
+    * Sections
+* **Comment styles**
+    * Recommend to add the “/** */” (Javadoc) style comments to all interface methods, with explanation of the parameters and returns. For example:
+
+      ```java
+          interface Transformation<T, R>
+          {
+              /**
+               * Processes input elements and returns current transformation state.
+               *
+               * @param elementOptional an element to be transformed. Will be empty
+               * when there are no more elements. In such case transformation should
+               * finish processing and flush any remaining data.
+               * @return the current transformation state, optionally bearing a result
+               * @see TransformationState#needsMoreData()
+               * @see TransformationState#blocked(ListenableFuture)
+               * @see TransformationState#yield()
+               * @see TransformationState#ofResult(Object)
+               * @see TransformationState#ofResult(Object, boolean)
+               * @see TransformationState#finished()
+               */
+              TransformationState<R> process(Optional<T> elementOptional);
+          }
+      ```
+
+    * Recommend to add the “/** */” style comments to important or difficult public methods, with explanation of the parameters and returns.
```

Comment:
```suggestion
    * Add the “/** */” style comments to important or difficult public methods, with explanation of the parameters and returns.
```
Same as above: if there's a good objection to the previous similar comment, it'll probably also be relevant here.